gcp_variant_transforms/beam_io/vcfio.py

```diff
@@ -207,6 +209,156 @@ def __repr__(self):
         [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
 
 
+
+class _VcfSink(filebasedsink.FileBasedSink):
+  """A sink for writing VCF files."""
+
+  class _ToVcfRecordCoder(coders.Coder):
+    """Coder for encoding :class:`Variant` objects as VCF text lines."""
+
+    def _is_alternate_phaseset(self, call):
+      return call.phaseset and call.phaseset != DEFAULT_PHASESET_VALUE
+
+    def _format_value(self, value):
+      """Formats a Variant field to be output to a VCF file."""
+      if not value and value != 0:
+        return MISSING_FIELD_VALUE
+      elif isinstance(value, list):
+        return ','.join([self._format_value(x) for x in value])
+      return str(value)
+
+    def _get_variant_info(self, variant):
+      info = []
+      # Set END in info if it doesn't match start+len(reference_bases). This is
+      # usually the case for non-variant regions.
+      if (variant.start is not None
+          and variant.reference_bases
+          and variant.end
+          and variant.start + len(variant.reference_bases) != variant.end):
+        info.append('END=%d' % variant.end)
+      # Set all other fields of info.
+      for k, v in variant.info.iteritems():
+        if v.data is True:
+          info.append(k)
+        else:
+          info.append('%s=%s' % (k, self._format_value(v.data)))
+      return info
+
+    def _get_variant_format_keys(self, variant):
+      if not variant.calls:
+        return []
+
+      format_keys = [GENOTYPE_FORMAT_KEY]
+      for call in variant.calls:
+        # If any calls have a set phaseset that is not `DEFAULT_PHASESET_VALUE`,
+        # the key will be added to the format field.
+        if self._is_alternate_phaseset(call):
+          format_keys.append(PHASESET_FORMAT_KEY)
+        format_keys.extend([k for k in call.info])
+
+      # Sort all keys and remove duplicates after GT
+      format_keys[1:] = sorted(list(set(format_keys[1:])))
+
+      return format_keys
+
+    def _get_variant_calls(self, variant, format_keys):
+      calls = []
+      for call in variant.calls:
+        call_info = []
+        # GT will always be the first format key.
+        genotype = []
+        for allele in call.genotype:
+          if allele == MISSING_GENOTYPE_VALUE:
+            genotype.append(MISSING_FIELD_VALUE)
+          else:
+            genotype.append(self._format_value(allele))
+        phase_char = '|' if call.phaseset else '/'
+        call_info.append(phase_char.join(genotype))
```

Comment:
there's the possibility of genotype itself being an empty list (it won't happen from our imports; but it's a reasonable thing to have). I think it would be fine if you just do: "call_info.append(phase_char.join(genotype) or MISSING_FIELD_VALUE)"

Reply:
Good catch!